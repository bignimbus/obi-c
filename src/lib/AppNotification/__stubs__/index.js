import AppNotification from '../';
import difference from 'lodash/difference';

const VALID_PROPS = [
  'body',
  'time',
  'title',
  'notifiable',
];

export const stubAppNotification = (props) => {
  const propKeys = Object.keys(props);
  const unrecognizedKeys = difference(propKeys, VALID_PROPS);
  if (unrecognizedKeys.length) {
    throw new Error(`invalid props: ${unrecognizedKeys.join(', ')}`);
  }
  return new AppNotification(props);
};

export const stubValidAppNotification = props => stubAppNotification({
  body: 'Bar',
  title: 'Foo',
  time: new Date(),
  notifiable: {},
  ...props,
});

export const stubAppNotificationWithInvalidBody = props => stubValidAppNotification({
  body: null,
  ...props,
});

export const stubAppNotificationWithInvalidTitle = props => stubValidAppNotification({
  title: null,
  ...props,
});

export const stubAppNotificationWithInvalidTime = props => stubValidAppNotification({
  time: null,
  ...props,
});
