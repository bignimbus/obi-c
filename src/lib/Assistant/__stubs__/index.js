import Assistant from '../';
import difference from 'lodash/difference';

const VALID_PROPS = [
  'hunger',
  'boredom',
];

export const stubAssistant = (props) => {
  const propKeys = Object.keys(props);
  const unrecognizedKeys = difference(propKeys, VALID_PROPS);
  if (unrecognizedKeys.length) {
    throw new Error(`invalid props: ${unrecognizedKeys.join(', ')}`);
  }
  return new Assistant(props);
};

export const stubValidAssistant = props => stubAssistant({ hunger: 50, boredom: 50, ...props });

export const stubStarvingAssistant = props => stubValidAssistant({ hunger: 100, ...props });
export const stubFullAssistant = props => stubValidAssistant({ hunger: 0, ...props });

export const stubBoredAssistant = props => stubValidAssistant({ boredom: 100, ...props });
export const stubFulfilledAssistant = props => stubValidAssistant({ boredom: 0, ...props });
