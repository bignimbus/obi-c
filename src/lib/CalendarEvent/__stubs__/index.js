import CalendarEvent from '../';
import difference from 'lodash/difference';

const VALID_PROPS = [
  'endTime',
  'startTime',
  'description',
];

export const stubCalendarEvent = (props) => {
  const propKeys = Object.keys(props);
  const unrecognizedKeys = difference(propKeys, VALID_PROPS);
  if (unrecognizedKeys.length) {
    throw new Error(`invalid props: ${unrecognizedKeys.join(', ')}`);
  }
  return new CalendarEvent(props);
};

export const stubValidCalendarEvent = () => stubCalendarEvent({
  endTime: new Date(),
  startTime: new Date(),
  description: 'A valid calendar event',
});

export const stubCalendarEventWithInvalidDate = () => stubCalendarEvent({
  endTime: null,
  startTime: null,
  description: null,
});
