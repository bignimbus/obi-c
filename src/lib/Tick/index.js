import Model from '../Model';

class Tick extends Model {
  init({
    user,
    assistant,
    incrementStart,
  }) {
    Object.assign(this, {
      user,
      assistant,
      incrementStart,
      incrementEnd: new Date(),
    });
  }

  getActiveNotifications () {
    return this.user.events.reduce((activeNotifications, { notifications }) => [
      ...activeNotifications,
      ...notifications.filter(
        ({ time }) => time >= this.incrementStart && time <= this.incrementEnd
      ),
    ], []);
  }
}

export default Tick;
