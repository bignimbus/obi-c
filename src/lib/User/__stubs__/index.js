import User from '../';
import difference from 'lodash/difference';

const VALID_PROPS = [
  'name',
  'events',
];

export const stubUser = (props) => {
  const propKeys = Object.keys(props);
  const unrecognizedKeys = difference(propKeys, VALID_PROPS);
  if (unrecognizedKeys.length) {
    throw new Error(`invalid props: ${unrecognizedKeys.join(', ')}`);
  }
  return new User(props);
};

export const stubValidUser = props => new User({
  name: 'Name',
  events: [],
  ...props,
});
