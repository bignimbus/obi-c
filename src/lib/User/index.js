import Model from '../Model';

class User extends Model {
  init ({
    id,
    name,
  }) {
    Object.assign(this, { id, name });
  }
}

export default User;
