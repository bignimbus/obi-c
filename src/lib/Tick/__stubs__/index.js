import difference from 'lodash/difference';
import Tick from '../';
import { stubValidUser } from '../../User/__stubs__';
import { stubValidAssistant } from '../../Assistant/__stubs__';
import { stubValidCalendarEventWithNotification } from '../../CalendarEvent/__stubs__';

export const minutesAgo = n => new Date(new Date().getTime() - n * 60 * 1000);

const VALID_PROPS = [
  'user',
  'assistant',
  'incrementStart',
];

export const stubTick = (props) => {
  const propKeys = Object.keys(props);
  const unrecognizedKeys = difference(propKeys, VALID_PROPS);
  if (unrecognizedKeys.length) {
    throw new Error(`invalid props: ${unrecognizedKeys.join(', ')}`);
  }
  return new Tick(props);
};

export const stubValidTick = props => stubTick({
  user: stubValidUser(),
  incrementStart: minutesAgo(1),
  assistant: stubValidAssistant(),
  ...props,
});

export const stubValidTickWithNotifications = (startTimes) => {
  const user = stubValidUser();
  const events = startTimes.map(t => stubValidCalendarEventWithNotification({
      user,
      owner: user,
      startTime: t,
      description: `Notification at ${t}`,
    }));
  user.events = events;
  return stubValidTick({ user });
};
