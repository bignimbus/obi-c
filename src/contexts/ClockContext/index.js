import React, {
  useState,
  useEffect,
  createContext,
} from 'react';
import PropTypes from 'prop-types';
import User from '../../lib/User';
import Tick from '../../lib/Tick';

const ClockContext = createContext();

export const ClockContextProvider = ({ children }) => {
  const [user] = useState(new User({}));
  const [tick, setTick] = useState(new Tick({
    user,
    incrementStart: new Date(),
  }));
  const [activeNotifications, setActiveNotifications] = useState(tick.getActiveNotifications());

  const incrementClock = () => {
    const interval = setInterval(() => {
      const { incrementEnd } = tick;
      const fifteenSecondsAfterLastTickInterval = new Date(incrementEnd.getTime() + 15 * 1000);
      const now = new Date();
      if (fifteenSecondsAfterLastTickInterval > now) return;
      const newTick = new Tick({ user, incrementStart: incrementEnd });
      setTick(newTick);
      setActiveNotifications(newTick.getActiveNotifications());
    }, 1000);
    return clearInterval.bind(null, interval);
  };

  useEffect(incrementClock);

  return (
    <ClockContext.Provider
      value={{
        activeNotifications,
        addEvent: user.addEvent.bind(user),
        removeEvent: user.removeEvent.bind(user),
      }}
    >
      { children }
    </ClockContext.Provider>
  );
};

ClockContextProvider.propTypes = {
  children: PropTypes.node.isRequired,
};

export default ClockContext;
