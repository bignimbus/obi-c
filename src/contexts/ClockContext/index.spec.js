import React, { useContext } from 'react';
import renderer, { act } from 'react-test-renderer';
import { advanceTo, clear } from 'jest-date-mock';
import ClockContext, { ClockContextProvider } from '.';
import { minutesAgo } from '../../lib/Tick/__stubs__';
import User from '../../lib/User';

const renderComponentWithContext = Child => renderer.create(
  <ClockContextProvider>
    <Child />
  </ClockContextProvider>,
);

describe('ClockContextProvider', () => {
  it('should expose a collection of notifications', () => {
    const TestComponent = () => {
      const { activeNotifications } = useContext(ClockContext);
      return <div>{ JSON.stringify(activeNotifications) } </div>;
    };
    const component = renderComponentWithContext(TestComponent);
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('should expose a function to add an event', () => {
    const spy = jest.spyOn(User.prototype, 'addEvent');
    const TestComponent = () => {
      const { addEvent } = useContext(ClockContext);
      const handleClick = () => {
        addEvent({
          description: 'Fake event',
          startTime: minutesAgo(-10),
        });
      };
      return (
        <button onClick={handleClick}>
          Add event
        </button>
      );
    };
    const component = renderComponentWithContext(TestComponent);
    component.root.findByType('button').props.onClick();
    expect(spy).toHaveBeenCalled();
  });

  it('should expose a function to remove an event', () => {
    const spy = jest.spyOn(User.prototype, 'removeEvent');
    const TestComponent = () => {
      const { removeEvent } = useContext(ClockContext);
      const handleClick = () => {
        removeEvent({
          description: 'Fake event',
          startTime: minutesAgo(10),
        });
      };
      return (
        <button onClick={handleClick}>
          Remove event
        </button>
      );
    };
    const component = renderComponentWithContext(TestComponent);
    component.root.findByType('button').props.onClick();
    expect(spy).toHaveBeenCalled();
  });

  it('should tick the clock every 15 seconds', async () => {
    jest.useFakeTimers();

    const TestComponent = () => {
      const { addEvent, activeNotifications } = useContext(ClockContext);

      const handleClick = (props) => {
        addEvent(props);
      };

      return (
        <div>
          <ul>
            {
              activeNotifications.map((n, i) => (
                <li key={i}>{ n.body }</li>
              ))
            }
          </ul>
          <button onClick={handleClick}>
            Add event
          </button>
        </div>
      );
    };
    const component = renderComponentWithContext(TestComponent);
    const { onClick } = component.root.findByType('button').props;
    onClick({ description: 'active right away', startTime: minutesAgo(2) });
    onClick({ description: 'active right away again', startTime: minutesAgo(1) });
    onClick({ description: 'active in a minute', startTime: minutesAgo(-1) });
    onClick({ description: 'active in an hour', startTime: minutesAgo(-60) });
    act(() => {
      advanceTo(minutesAgo(-16 / 60));
      jest.advanceTimersByTime(16 * 1000);
    });
    expect(component.toJSON()).toMatchSnapshot();
    act(() => {
      advanceTo(minutesAgo(-1));
      jest.advanceTimersByTime(60 * 1000);
    });
    expect(component.toJSON()).toMatchSnapshot();
    act(() => {
      advanceTo(minutesAgo(-60));
      jest.advanceTimersByTime(60 * 60 * 1000);
    });
    expect(component.toJSON()).toMatchSnapshot();
  });
  clear();
});
