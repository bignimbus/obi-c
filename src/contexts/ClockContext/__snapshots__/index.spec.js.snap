// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClockContextProvider should expose a collection of notifications 1`] = `
<div>
  []
   
</div>
`;

exports[`ClockContextProvider should tick the clock every 15 seconds 1`] = `
<div>
  <ul>
    <li>
      active right away
    </li>
    <li>
      active right away again
    </li>
  </ul>
  <button
    onClick={[Function]}
  >
    Add event
  </button>
</div>
`;

exports[`ClockContextProvider should tick the clock every 15 seconds 2`] = `
<div>
  <ul>
    <li>
      active right away
    </li>
    <li>
      active right away again
    </li>
    <li>
      active in a minute
    </li>
  </ul>
  <button
    onClick={[Function]}
  >
    Add event
  </button>
</div>
`;

exports[`ClockContextProvider should tick the clock every 15 seconds 3`] = `
<div>
  <ul>
    <li>
      active right away
    </li>
    <li>
      active right away again
    </li>
    <li>
      active in a minute
    </li>
    <li>
      active in an hour
    </li>
  </ul>
  <button
    onClick={[Function]}
  >
    Add event
  </button>
</div>
`;
